tests/core/featurizers/test_precomputation.py

```diff
@@ -0,0 +1,565 @@
+        sub_states.append(sub_state_copy)
+
+    # if we handle the collisions...
+    lookup_table = CoreFeaturizationPrecomputations(handle_collisions=True)
+    for sub_state in sub_states:
+        lookup_table.add(Message(sub_state))
+    half_of_added_messages = int(len(sub_states) / 2)
+    assert len(lookup_table) == half_of_added_messages
+    for message in lookup_table.values():
+        assert ARBITRARY_NON_KEY in message.data
+    assert lookup_table.num_collisions_ignored == 0
+    assert lookup_table.num_collisions_resolved == half_of_added_messages
+
+    # ... and if we don't
+    lookup_table2 = CoreFeaturizationPrecomputations(handle_collisions=False)
+    for idx in range(0, half_of_added_messages):
+        lookup_table2.add(Message(sub_states[idx]))
+    for idx in range(half_of_added_messages, 2 * half_of_added_messages):
+        with pytest.raises(ValueError, match="Expected added message to be consistent"):
+            lookup_table2.add(Message(sub_states[idx]))
+
+
+def test_precomputation_build_with_inconsistencies_due_to_missing_features():
+    # This really should not happen, but if it did...
+
+    # construct a set of unique substates and messages
+    dummy_value = "this-could-be-anything"
+    unique_sub_states = [
+        {INTENT: "greet"},
+        {TEXT: "text", ENTITIES: dummy_value},
+        {ACTION_TEXT: "action_text"},
+        {ACTION_NAME: "action_name"},
+    ]
+
+    # add different numbers of features
+    dummy_feature = Features(
+        np.ndarray((1, 2)),
+        attribute="this-is-never-checked-and-can-be-anything",
+        feature_type="really-anything",
+        origin="",
+    )
+    messages = []
+    for sub_state in unique_sub_states:
+        messages.append(Message(data=sub_state, features=[]))
+        messages.append(Message(data=sub_state, features=[dummy_feature]))
+        messages.append(
+            Message(data=sub_state, features=[dummy_feature, dummy_feature])
+        )
+
+    # if we do collision handling...
+    lookup_table = CoreFeaturizationPrecomputations(handle_collisions=True)
+    for message in messages:
+        lookup_table.add(message)
+    assert len(lookup_table) == len(unique_sub_states)
+    for message in lookup_table.values():
+        assert message.features is not None
+        assert len(message.features) == 2
+    assert lookup_table.num_collisions_ignored == 0
+    assert lookup_table.num_collisions_resolved == len(unique_sub_states) * 2
+
+    # ... and if we don't
+    lookup_table2 = CoreFeaturizationPrecomputations(handle_collisions=False)
+    lookup_table2.add(messages[0])
+    with pytest.raises(ValueError, match="Expected added message to be consistent"):
+        lookup_table2.add(messages[1])
+
+
+def test_precomputation_build_with_inconsistent_data():
+    """Tests whether the lookup table can deal not-too messy training data.
+    That is, we assume that someone missed to annotate `ENTITIES`
```

Comment:
```suggestion
    That is, we assume that someone missed to annotate `ENTITIES`.
```

Reply:
Was thinking of story data where you annotate two instances of the same text differently - that would be lost at the moment - but that a) should/could be handled differently then and b) I think I can remove entities and so this is not really an issue